# Subject: Classical Data Analysis

## Session 1 - Regression

### Individual assignment 1

Develop a regression analysis in Statmodels (with and without a constant) and SKLearn, based on the Iris sklearn dataset. This data sets consists of 3 different types of irises’ (Setosa, Versicolour, and Virginica) petal and sepal length.

See here for more information on this dataset: https://en.wikipedia.org/wiki/Iris_flower_data_set 

Use the field “sepal width (cm)” as independent variable and the field “sepal length (cm)” as dependent variable.

- Interpret and discuss the OLS Regression Results.
- Commit scripts in your GitHub account. You should export your solution code (.ipynb notebook) and push it to your repository “ClassicalDataAnalysis”.

The following are the tasks that should complete and synchronize with your repository “ClassicalDataAnalysis” until October 13. Please notice that none of these tasks is graded, however it’s important that you correctly understand and complete them in order to be sure that you won’t have problems with further assignments.

# Linear Regression in Statsmodels

## Load the iris dataset

In [68]:
from sklearn import linear_model
from sklearn import datasets
import matplotlib.pyplot as plt
import numpy as np
import statsmodels.api as sm
import pandas as pd

In [69]:
iris = datasets.load_iris()

In [70]:
# why cant i do: iris.head()????

In [71]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

### Regression model with Statsmodels and without a constant:

In [72]:
df = pd.DataFrame(iris.data, columns=iris.feature_names) 

In [73]:
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [74]:
X = df["sepal width (cm)"]
y = df["sepal length (cm)"]

In [75]:
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:      sepal length (cm)   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     3316.
Date:                Tue, 17 Oct 2017   Prob (F-statistic):          1.04e-103
Time:                        23:57:09   Log-Likelihood:                -243.13
No. Observations:                 150   AIC:                             488.3
Df Residuals:                     149   BIC:                             491.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
sepal width (cm)     1.8717      0.033     57.585      0.000       1.807       1.936
==============================================================================
Omnibus:                       16.884   Durbin-Watson:                   0.429
Prob(Omnibus):                  0.000   Jarque-Bera (JB):                7.669
Skew:                          -0.336   Prob(JB):                       0.0216
Kurtosis:                       2.120   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpreting the Table 

The coefficient of 1.8717 means that as sepal width increases by 1cm, sepal length increases by 1.8717cm. R-squared is only 0.957 means the model is a very good model

### Regression model with Statsmodels and with a constant:

In [76]:
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:      sepal length (cm)   R-squared:                       0.012
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     1.792
Date:                Tue, 17 Oct 2017   Prob (F-statistic):              0.183
Time:                        23:57:11   Log-Likelihood:                -183.14
No. Observations:                 150   AIC:                             370.3
Df Residuals:                     148   BIC:                             376.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const                6.4812      0.481     13.466      0.000       5.530       7.432
sepal width (cm)    -0.2089      0.156     -1.339      0.183      -0.517       0.099
==============================================================================
Omnibus:                        4.455   Durbin-Watson:                   0.941
Prob(Omnibus):                  0.108   Jarque-Bera (JB):                4.252
Skew:                           0.356   Prob(JB):                        0.119
Kurtosis:                       2.585   Cond. No.                         24.3
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpreting the Table 

We have now a sepal width-intercept at -0.2089. We also changed the slope of the RM predictor from 1.8717 to 6.4812. R-squared decreased from 0.957 to 0.012, indicating a very bad model with a constant.

# Linear Regression in SKLearn 

In [77]:
from sklearn import linear_model
from sklearn import datasets

iris = datasets.load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names) 

In [78]:
df["sepal width (cm)"]

0      3.5
1      3.0
2      3.2
3      3.1
4      3.6
5      3.9
6      3.4
7      3.4
8      2.9
9      3.1
10     3.7
11     3.4
12     3.0
13     3.0
14     4.0
15     4.4
16     3.9
17     3.5
18     3.8
19     3.8
20     3.4
21     3.7
22     3.6
23     3.3
24     3.4
25     3.0
26     3.4
27     3.5
28     3.4
29     3.2
      ... 
120    3.2
121    2.8
122    2.8
123    2.7
124    3.3
125    3.2
126    2.8
127    3.0
128    2.8
129    3.0
130    2.8
131    3.8
132    2.8
133    2.8
134    2.6
135    3.0
136    3.4
137    3.1
138    3.0
139    3.1
140    3.1
141    3.1
142    2.7
143    3.2
144    3.3
145    3.0
146    2.5
147    3.0
148    3.4
149    3.0
Name: sepal width (cm), Length: 150, dtype: float64

In [79]:
X = df["sepal width (cm)"]
y = df["sepal length (cm)"]

In [80]:
lm = linear_model.LinearRegression()

In [81]:
model = lm.fit(X,y)
predictions = lm.predict(X)
# dont knwo how to solve this error...

Y:\BTS\Program\lib\site-packages\sklearn\utils\validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)


ValueError: Found input variables with inconsistent numbers of samples: [1, 150]